web/srv/api_handlers.go

```diff
@@ -242,3 +246,39 @@ func (h *handler) handleApiTap(w http.ResponseWriter, req *http.Request, p httpr
 		}
 	}
 }
+
+func (h *handler) handleProfileDownload(w http.ResponseWriter, req *http.Request, params httprouter.Params) {
+	service := req.FormValue("service")
+	namespace := req.FormValue("namespace")
+
+	if service == "" || namespace == "" {
+		err := fmt.Errorf("Service and namespace must be provided to create a new profile")
+		log.Error(err.Error())
+		renderJsonError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	tmplParams := profileCmd.BuildConfig(namespace, service)
+
+	template, err := template.New("profile").Parse(profile.Template)
+	if err != nil {
+		log.Error(err.Error())
+		renderJsonError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	profileYaml := &bytes.Buffer{}
+	err = template.Execute(profileYaml, tmplParams)
+	if err != nil {
+		log.Error(err.Error())
+		renderJsonError(w, err, http.StatusInternalServerError)
+		return
+	}
```

Comment:
You could reuse more of the code from `cli/cmd/profile.go` if you do something like:

```diff
diff --git a/cli/cmd/profile.go b/cli/cmd/profile.go
index 3b1a1330..3123ab9e 100644
--- a/cli/cmd/profile.go
+++ b/cli/cmd/profile.go
@@ -54,7 +54,7 @@ func newCmdProfile() *cobra.Command {
 				return errors.New("only template mode is currently supported, please run with --template")
 			}
 
-			return renderProfileTemplate(BuildConfig(options.namespace, args[0]), os.Stdout)
+			return RenderProfileTemplate(BuildConfig(options.namespace, args[0]), os.Stdout)
 		},
 	}
 
@@ -73,7 +73,7 @@ func BuildConfig(namespace, service string) *ProfileTemplateConfig {
 	}
 }
 
-func renderProfileTemplate(config *ProfileTemplateConfig, w io.Writer) error {
+func RenderProfileTemplate(config *ProfileTemplateConfig, w io.Writer) error {
 	template, err := template.New("profile").Parse(profile.Template)
 	if err != nil {
 		return err
diff --git a/web/srv/api_handlers.go b/web/srv/api_handlers.go
index 84a18829..1194f757 100644
--- a/web/srv/api_handlers.go
+++ b/web/srv/api_handlers.go
@@ -6,7 +6,6 @@ import (
 	"fmt"
 	"io"
 	"net/http"
-	"text/template"
 	"time"
 
 	"github.com/golang/protobuf/jsonpb"
@@ -14,7 +13,6 @@ import (
 	"github.com/gorilla/websocket"
 	"github.com/julienschmidt/httprouter"
 	profileCmd "github.com/linkerd/linkerd2/cli/cmd"
-	"github.com/linkerd/linkerd2/cli/profile"
 	"github.com/linkerd/linkerd2/controller/api/util"
 	pb "github.com/linkerd/linkerd2/controller/gen/public"
 	"github.com/linkerd/linkerd2/pkg/k8s"
@@ -259,19 +257,12 @@ func (h *handler) handleProfileDownload(w http.ResponseWriter, req *http.Request
 	}
 
 	tmplParams := profileCmd.BuildConfig(namespace, service)
-
-	template, err := template.New("profile").Parse(profile.Template)
-	if err != nil {
-		log.Error(err.Error())
-		renderJsonError(w, err, http.StatusInternalServerError)
-		return
-	}
-
 	profileYaml := &bytes.Buffer{}
-	err = template.Execute(profileYaml, tmplParams)
+
+	err := profileCmd.RenderProfileTemplate(tmplParams, profileYaml)
 	if err != nil {
 		log.Error(err.Error())
-		renderJsonError(w, err, http.StatusInternalServerError)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
```

(you'd want to move that `RenderProfileTemplate` func to somewhere in `pkg` though, per my other suggestion)